crates/pathfinder/src/state/block_hash.rs

```diff
@@ -0,0 +1,310 @@
+use anyhow::{Context, Error, Result};
+use bitvec::prelude::BitView;
+use stark_hash::{stark_hash, StarkHash};
+
+use crate::core::{SequencerAddress, StarknetBlockHash};
+use crate::sequencer::reply::{
+    transaction::{Event, Receipt, Transaction},
+    Block,
+};
+use crate::state::merkle_tree::MerkleTree;
+
+pub fn compute_block_hash(block: &Block) -> Result<StarknetBlockHash> {
+    let transaction_commitment = calculate_transaction_commitment(&block.transactions)?;
+    let event_commitment = calculate_event_commitment(&block.transaction_receipts)?;
+
+    anyhow::ensure!(block.block_number.is_some());
+    let block_number = block.block_number.unwrap();
+    anyhow::ensure!(block.state_root.is_some());
+    let state_root = block.state_root.unwrap();
+
+    let num_transactions: u64 = block.transactions.len().try_into()?;
+    let num_events = number_of_events_in_block(block);
+    let num_events: u64 = num_events.try_into()?;
+
+    let sequencer_address = block
+        .sequencer_address
+        .unwrap_or(SequencerAddress(StarkHash::ZERO));
+
+    let data = [
+        // block number
+        StarkHash::from_u64(block_number.0),
+        // global state root
+        state_root.0,
+        // sequencer address
+        sequencer_address.0,
+        // block timestamp
+        StarkHash::from_u64(block.timestamp.0),
+        // number of transactions
+        StarkHash::from_u64(num_transactions),
+        // transaction commitment
+        transaction_commitment,
+        // number of events
+        StarkHash::from_u64(num_events),
+        // event commitment
+        event_commitment,
+        // reserved: protocol version
+        StarkHash::ZERO,
+        // reserved: extra data
+        StarkHash::ZERO,
+        // parent block hash
+        block.parent_block_hash.0,
+    ];
+
+    let block_hash = stark_hash_of_array(data.into_iter());
+
+    Ok(StarknetBlockHash(block_hash))
+}
+
+#[derive(Default)]
+struct CommitmentTree {
+    tree: MerkleTree<()>,
```

Comment:
```suggestion
    tree: MerkleTree<NoPersistance>,
```
Should the null storage maybe get a proper type? Its confusing to read if you don't know the empty tuple's impl.

Reply:
Rationale for that was .... unit cannot really have any state so it's implementation would be obvious. But this is here an impl detail.